### PR TITLE
New Feature: Support User-provided Mix Spatial Mapping, And spatial_mapping_hint Definition Can Be Provided in the Mapping File for Mix Spatial Mapping Auto-generation.

### PR DESCRIPTION
**Pull request by JiacongSun**
### New feature: Supporting Mixed Spatial Mapping
We have introduced a new feature that enables the inclusion of mixed spatial mapping in the mapping file, alongside the original single-layer-dimension spatial mapping. For examples, please refer to the files located under the `/tests/main/test_with_mix_spatial_mapping` folder.

### New parameter in the input mapping file:
spatial_mapping_hint: `Dict: {hardware_dimension: [supported_layer_dimensions]}`. This parameter only works when `spatial_mapping `is not provided or partially provided in the mapping file. Now the generated spatial mappings will only be on the layer dimensions defined in spatial_mapping_hint. It will be assumed to include all layer dimensions if this parameter is not provided.

### New parameters can be provided from the top:
`enable_mix_spatial_mapping_generation`: 

- **True** --> `SpatialMappingGeneratorStage `will auto-generate mix spatial mapping. 
- **False** --> `SpatialMappingGeneratorStage `will stick to spatial mapping of single layer dimension.

`maximize_hardware_utilization`: 

- **True** --> `SpatialMappingGeneratorStage `will only keep two spatial mapping candidates for each hardware dimension, which has the highest hardware utilization, for a faster simulation time. 
- **False** --> `SpatialMappingGeneratorStage `will keep all spatial mapping candidates it found.

### Adds/Changes:

- `SpatialMappingGeneratorStage `now can also generate mix spatial mappings.
- `SpatialMappingConversionStage `now can also interpret mix spatial mappings.
- Pytest cases are added to the `/tests/main/test_with_mix_spatial_mapping` folder for testing scenarios involving mixed spatial mapping. Additionally, a dedicated test case with a `spatial_mapping_hint `definition can be found in `test_tesla_npu_like.py`.

### Bug Fixes:

- The abortion in `SpatialMappingGeneratorStage `has been fixed for cases where only parts of the spatial mapping are user-provided.
- An error in `LomaStage`, where the same layer dimension was mapped to different hardware dimensions, has been resolved. Now, these spatial mappings on different hardware dimensions are merged before being sent to `LomaStage`.

### CODE REVIEW

File: zigzag/api.py [hunk 1]
```diff
@@ -231,6 +231,84 @@ def get_hardware_performance_zigzag_without_unused_memory(
 
     return cmes[0][0].energy_total, cmes[0][0].latency_total2, cmes
 
+def get_hardware_performance_zigzag_with_mix_spatial_mapping(
```
Comment (asyms) [inline zigzag/api.py:234]: How is this API call different from the other ones? If there are no differences we should avoid making new api calls

File: tests/main/test_with_mix_spatial_mapping/test_edge_tpu_like.py [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+
+
+@pytest.fixture
+def accelerator():
+    return "zigzag.inputs.examples.hardware.Edge_TPU_like"
+
+
+@pytest.mark.parametrize("workload", workloads)
+def test_api(workload, accelerator, mapping):
+    (energy, latency, cmes) = get_hardware_performance_zigzag_without_unused_memory(
```
Comment (asyms) [inline tests/main/test_with_mix_spatial_mapping/test_edge_tpu_like.py]: Wrong api call?

File: tests/main/test_with_mix_spatial_mapping/test_meta_prototype_like.py [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+
+
+@pytest.fixture
+def accelerator():
+    return "zigzag.inputs.examples.hardware.Meta_prototype"
+
+
+@pytest.mark.parametrize("workload", workloads)
+def test_api(workload, accelerator, mapping):
+    (energy, latency, cmes) = get_hardware_performance_zigzag_without_unused_memory(
```
Comment (asyms) [inline tests/main/test_with_mix_spatial_mapping/test_meta_prototype_like.py]: Wrong api call?

File: tests/main/test_with_mix_spatial_mapping/test_tesla_npu_like.py [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+
+
+@pytest.fixture
+def accelerator():
+    return "zigzag.inputs.examples.hardware.Tesla_NPU_like"
+
+
+@pytest.mark.parametrize("workload", workloads)
+def test_api(workload, accelerator, mapping):
+    (energy, latency, cmes) = get_hardware_performance_zigzag_without_unused_memory(
```
Comment (asyms) [inline tests/main/test_with_mix_spatial_mapping/test_tesla_npu_like.py]: Wrong api call?

File: tests/main/test_with_mix_spatial_mapping/test_tpu_like.py [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+
+
+@pytest.fixture
+def accelerator():
+    return "zigzag.inputs.examples.hardware.TPU_like"
+
+
+@pytest.mark.parametrize("workload", workloads)
+def test_api(workload, accelerator, mapping):
+    (energy, latency, cmes) = get_hardware_performance_zigzag_without_unused_memory(
```
Comment (asyms) [inline tests/main/test_with_mix_spatial_mapping/test_tpu_like.py]: Wrong api call?

File: zigzag/api.py (hunk 1 above)
Comment (asyms) [reply zigzag/api.py:234]: I suppose only the last two parameters given to the MainStage are new. This api call is fine then.

**Review by asyms — CHANGES_REQUESTED** [review]
Please use the correct API call and add tests for the following edge cases:

1. No hint provided
2. Hint provided but it's incompatible with the architecture
3. Too many dimensions provided in hint

File: zigzag/api.py (hunk 1 above)
Comment (JiacongSun) [reply zigzag/api.py:234]: Yes, that's correct. Only the last two parameters are different. All the others are the same to function `get_hardware_performance_zigzag_without_unused_memory`.

File: tests/main/test_with_mix_spatial_mapping/test_edge_tpu_like.py (hunk 2 above)
Comment (JiacongSun) [reply tests/main/test_with_mix_spatial_mapping/test_edge_tpu_like.py]: Great review! Fixing.

File: tests/main/test_with_mix_spatial_mapping/test_meta_prototype_like.py (hunk 3 above)
Comment (JiacongSun) [reply tests/main/test_with_mix_spatial_mapping/test_meta_prototype_like.py]: Great review! Fixing.

File: tests/main/test_with_mix_spatial_mapping/test_tesla_npu_like.py (hunk 4 above)
Comment (JiacongSun) [reply tests/main/test_with_mix_spatial_mapping/test_tesla_npu_like.py]: Great review! Fixing.

File: tests/main/test_with_mix_spatial_mapping/test_tpu_like.py (hunk 5 above)
Comment (JiacongSun) [reply tests/main/test_with_mix_spatial_mapping/test_tpu_like.py]: Great review! Fixing.

**Comment by JiacongSun** [comment]
### New fixes:

- Corrected an erroneous API call in the original pytest cases.

- Updated pytest cases to include five tests, each addressing different scenarios related to mixed spatial mapping features:

1. When only `spatial_mapping` is provided while `spatial_mapping_hint` is not provided (Test: `test_with_mix_spatial_mapping/test_ascend_like.py`).
2. When both `spatial_mapping` and `spatial_mapping_hint` are provided (Test: `test_with_mix_spatial_mapping/test_edge_tpu_like.py`).
3. When an incomplete `spatial_mapping` is provided, and `spatial_mapping_hint` is provided (Test: `test_with_mix_spatial_mapping/test_meta_prototype_like.py`).
4. When only `spatial_mapping_hint` is provided (Test: `test_with_mix_spatial_mapping/test_tesla_npu_like.py`).
5. When the number of dimensions provided in the `spatial_mapping_hint` exceeds the dimensions available in the hardware (Test: `test_with_mix_spatial_mapping/test_tpu_like.py`).

**Review by asyms — APPROVED** [review]
LGTM!
